runtime/vm/ContinuationHelpers.cpp

```diff
@@ -265,28 +289,69 @@ yieldContinuation(J9VMThread *currentThread, BOOLEAN isFinished)
 }
 
 void
-freeContinuation(J9VMThread *currentThread, j9object_t continuationObject)
+freeContinuation(J9VMThread *currentThread, j9object_t continuationObject, BOOLEAN isGlobal)
 {
 	J9VMContinuation *continuation = J9VMJDKINTERNALVMCONTINUATION_VMREF(currentThread, continuationObject);
 	if (NULL != continuation) {
-		PORT_ACCESS_FROM_VMC(currentThread);
-		/* Free all stack used by continuation */
-		J9JavaStack *currentStack = continuation->stackObject;
-		do {
+		ContinuationState continuationState = *VM_ContinuationHelpers::getContinuationStateAddress(currentThread, continuationObject);
+		Assert_VM_true(!VM_ContinuationHelpers::isConcurrentlyScanned(continuationState) && (NULL == VM_ContinuationHelpers::getCarrierThread(continuationState)));
+
+		/* Free old stack used by continuation */
+		J9JavaStack *currentStack = continuation->stackObject->previous;
+		while (NULL != currentStack) {
 			J9JavaStack *previous = currentStack->previous;
 
 			freeJavaStack(currentThread->javaVM, currentStack);
 			currentStack = previous;
-		} while (NULL != currentStack);
-
-		ContinuationState continuationState = *VM_ContinuationHelpers::getContinuationStateAddress(currentThread, continuationObject);
-		Assert_VM_true(!VM_ContinuationHelpers::isConcurrentlyScanned(continuationState) && (NULL == VM_ContinuationHelpers::getCarrierThread(continuationState)));
-
-		/* Free the J9VMContinuation struct */
-		j9mem_free_memory(continuation);
+		}
 
 		/* Update Continuation object's vmRef field */
 		J9VMJDKINTERNALVMCONTINUATION_SET_VMREF(currentThread, continuationObject, NULL);
+
+		recycleContinuation(currentThread->javaVM, currentThread, continuation, isGlobal);
+	}
+}
+
+void
+recycleContinuation(J9JavaVM *vm, J9VMThread *vmThread, J9VMContinuation* continuation, BOOLEAN isGlobal)
+{
+	/* Prepare continuationState for recycle */
+	/* Reset stack initial state */
+	J9SFJNINativeMethodFrame *frame = ((J9SFJNINativeMethodFrame*)continuation->stackObject->end) - 1;
+	frame->method = NULL;
+	frame->specialFrameFlags = 0;
+	frame->savedCP = NULL;
+	frame->savedPC = (U_8*)(UDATA)J9SF_FRAME_TYPE_END_OF_STACK;
+	frame->savedA0 = (UDATA*)(UDATA)J9SF_A0_INVISIBLE_TAG;
+	continuation->sp = (UDATA*)frame;
+	continuation->literals = NULL;
+	continuation->pc = (U_8*)J9SF_FRAME_TYPE_JNI_NATIVE_METHOD;
+	continuation->arg0EA = (UDATA*)&frame->savedA0;
+
+	if (!isGlobal && (NULL == vmThread->cachedContinuation)) {
+		/* If called by carrier thread (not global), try to store in local cache first. */
+		vmThread->cachedContinuation = continuation;
+	} else {
+		BOOLEAN cached = FALSE;
+		/* Greedily try to cache continuation struct in global array. */
+		for (U_32 i = 0; i < vm->continuationArraySize; i++) {
+			if ((NULL == vm->globalContinuationCacheArray[i])
+			&& (NULL == (UDATA*)VM_AtomicSupport::lockCompareExchange(
```

Comment:
Locally tested with an atomic counter to avoid looping the array when empty/full during create/recycle. no perf delta seen from test (Skynet). Deferring optimization on this area to future work